context.go

```diff
@@ -2,107 +2,154 @@ package parens
-	val, found := frame.vars[name]
-	return val, found
+// Context encapsulates execution state.  It is passed to Evaluator in order to evaluate
+// a form within a given execution context.
+type Context struct {
+	stack    []StackFrame
+	newMap   MapFactory
+	maxDepth int
 }
 
-// SetVar in the stack frame
-func (frame *StackFrame) SetVar(name string, val value.Any) {
-	if frame.vars == nil {
-		frame.vars = map[string]value.Any{}
+// NewContext returns a global context using the default Context implementation.
+// The returned Context IS NOT safe for concurrent use without external synchronization.
+func NewContext(opt ...Option) (ctx Context) {
+	for _, f := range withDefaults(opt) {
+		f(&ctx)
 	}
-	frame.vars[name] = val
+
+	globalFrame := StackFrame{Name: "<global>", ConcurrentMap: ctx.newMap()}
+	ctx.stack = append(ctx.stack, globalFrame)
+	return
 }
 
-type basicContext struct {
-	stack    []StackFrame
-	maxDepth int
+// NewFrame creates an empty StackFrame.  The ConcurrentMap is guaranteed to be non-nil
+// and ready for use.  Callers must populate the frame before pushing it to the stack.
+func (ctx Context) NewFrame() StackFrame {
+	return StackFrame{ConcurrentMap: ctx.newMap()}
 }
 
-// TODO(enhancement): max stack depth?
-func newContext(stack []StackFrame) *basicContext {
-	return &basicContext{
-		stack: stack,
+// Push a frame onto the stack
+func (ctx *Context) Push(frame StackFrame) {
+	if frame.ConcurrentMap == nil {
+		panic("ConcurrentMap must not be nil")
+	}
+
+	if len(ctx.stack) == ctx.maxDepth {
+		panic(fmt.Sprintf("stack-limit exceeded (maxdepth=%d)", ctx.maxDepth))
 	}
-}
 
-func (ctx *basicContext) Push(frame StackFrame) {
 	ctx.stack = append(ctx.stack, frame)
 }
 
-func (ctx *basicContext) Pop() (f StackFrame) {
+// Pop a frame from the stack
+func (ctx *Context) Pop() (f StackFrame) {
 	if len(ctx.stack) == 0 {
-		panic("Evaluator stack must never be empty")
+		panic("stack must never be empty")
 	}
 
 	f, ctx.stack = ctx.stack[len(ctx.stack)-1], ctx.stack[:len(ctx.stack)-1]
 	return
 }
 
-func (ctx basicContext) SetGlobal(name string, val value.Any) {
-	ctx.stack[0].SetVar(name, val)
+// SetGlobal variable
+func (ctx Context) SetGlobal(name string, val value.Any) {
+	ctx.stack[0].Store(name, val)
 }
 
-func (ctx *basicContext) Resolve(name string) (value.Any, error) {
+// Resolve name.
+func (ctx Context) Resolve(name string) (value.Any, error) {
 	if len(ctx.stack) == 0 {
 		panic("runtime stack must never be empty")
 	}
 
 	// traverse from top of the stack until a binding is found.
 	for i := len(ctx.stack) - 1; i >= 0; i-- {
-		if v, found := ctx.stack[i].GetVar(name); found {
+		if v, found := ctx.stack[i].Load(name); found {
 			return v, nil
 		}
 	}
 
 	return nil, ErrNotFound
 }
 
-func (ctx basicContext) NewChild() Context {
+// NewChild derives a copy of the Context
+func (ctx Context) NewChild() Context {
+	// TODO(performance):  Over-allocate by some reasonable ammount (8-16?).  NewChild is
```

Comment:
since we are using slices which uses array doubling, Go would already be over-allocating.. `cap > size`

Reply:
Oh cool, I didn't realize the compiler was that smart!